core/kernel/tee_ta_manager.c

```diff
@@ -823,6 +823,168 @@ TEE_Result tee_ta_invoke_command(TEE_ErrorOrigin *err,
 	return res;
 }
 
+#if defined(CFG_TA_STATS)
+static TEE_Result tee_ta_dump_memstats(struct tee_ta_session *s,
+				       uint32_t cancel_req_to,
+				       struct tee_ta_param *param)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_ta_ctx *ctx = NULL;
+	struct ts_ctx *ts_ctx = NULL;
+
+	ts_ctx = s->ts_sess.ctx;
+	if (ts_ctx)
+		ctx = ts_to_ta_ctx(ts_ctx);
+
+	if (ctx->panicked)
+		return TEE_ERROR_TARGET_DEAD;
+
+	if (tee_ta_try_set_busy(ctx)) {
+		s->param = param;
+		set_invoke_timeout(s, cancel_req_to);
+		res = ts_ctx->ops->dump_mem_stats(&s->ts_sess);
+		s->param = NULL;
+		tee_ta_clear_busy(ctx);
+	} else {
+		/* Deadlock avoided */
+		res = TEE_ERROR_BUSY;
+	}
+
+	return res;
+}
+
+TEE_Result tee_ta_instance_stats(void *buff,
+				 uint32_t *buff_size)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_ta_param param = { };
+	struct tee_ta_dump_stats *dump_stats = NULL;
+	struct tee_ta_session *sess = NULL;
+	struct tee_ta_session_head *open_sessions = NULL;
+	struct tee_ta_ctx *ctx = NULL;
+	uint32_t count = 0;
+	uint32_t size = 0;
+	uint32_t actual_size = 0;
+	uint32_t stats_num = 0;
+
+	if (!buff_size)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mutex_lock(&tee_ta_mutex);
+
+	/*
+	 * Go through all available TA and calc out the actual buffer size.
+	 */
+	TAILQ_FOREACH(ctx, &tee_ctxes, link)
+		actual_size += sizeof(struct tee_ta_dump_stats);
+
+	if (!buff || *buff_size == 0) {
+		/*
+		 * Buff is null or pass size with 0 means caller want to get
+		 * the size of buff.
+		 */
+		*buff_size = actual_size;
+		mutex_unlock(&tee_ta_mutex);
+		return res;
+	}
+
+	/*
+	 * If the buffer size is smaller than the actual size,
+	 * update *buff_size and return TEE_ERROR_SHORT_BUFFER
+	 */
+	size = *buff_size;
+	if (size < actual_size) {
+		*buff_size = actual_size;
+		res = TEE_ERROR_SHORT_BUFFER;
+		mutex_unlock(&tee_ta_mutex);
+		return res;
+	}
+
+	dump_stats = (struct tee_ta_dump_stats *)buff;
+	nsec_sessions_list_head(&open_sessions);
+	/*
+	 * Scan all sessions opened from secure side by searching through
+	 * all available TA instances and for each context, scan all opened
+	 * sessions.
+	 */
+	/* Context of a loaded TA */
+	TAILQ_FOREACH(ctx, &tee_ctxes, link) {
+		memcpy(&dump_stats->uuid, &ctx->ts_ctx.uuid,
+		       sizeof(ctx->ts_ctx.uuid));
+		dump_stats->flags = ctx->flags;
```

Comment:
Where are the checks to see that `UTEE_ENTRY_FUNC_DUMP_MEMSTATS` is supported so we don't panic the TA?

If we're not going to have such a flag the very least we can do is to change `__utee_entry()` to return an error instead panicking on unknown entry function ids. That should go in a separate patch before this patch then.

Reply:
> If we're not going to have such a flag the very least we can do is to change `__utee_entry()` to return an error instead panicking on unknown entry function ids. That should go in a separate patch before this patch then.

IMO that would be good enough. No need for a flag.

Reply:
will add a commit in this PR to cover this.